lib/ssz_ex/decode.ex

```diff
@@ -56,115 +55,137 @@ defmodule SszEx.Decode do
   end
 
-  defp decode_uint(_binary, size), do: {:error, "invalid byte size #{inspect(size)}"}
+  defp decode_uint(binary, size),
+    do:
+      {:error,
+       "Invalid binary length while decoding uint.\nExpected size: #{size}.\nFound:#{bit_size(binary)}\nBinary:#{inspect(binary)}"}
 
   defp decode_bool("\x01"), do: {:ok, true}
   defp decode_bool("\x00"), do: {:ok, false}
-  defp decode_bool(_), do: {:error, "invalid bool value"}
 
-  defp decode_bitlist(bit_list, max_size) when bit_size(bit_list) > 0 do
+  defp decode_bool(binary),
+    do:
+      {:error,
+       "Invalid binary value while decoding bool.\nExpected value: x01/x00.\nFound: x#{Base.encode16(binary)}."}
+
+  defp decode_bitlist(bit_list, _max_size) when bit_size(bit_list) == 0,
+    do: {:error, "Invalid binary value while decoding BitList.\nEmpty binary found.\n"}
+
+  defp decode_bitlist(bit_list, max_size) do
     num_bytes = byte_size(bit_list)
     decoded = BitList.new(bit_list)
     len = BitList.length(decoded)
 
     cond do
       match?(<<_::binary-size(num_bytes - 1), 0>>, bit_list) ->
-        {:error, "BitList has no length information."}
-
-      div(len, @bits_per_byte) + 1 != num_bytes ->
-        {:error, "invalid byte count"}
+        {:error, "Invalid binary value while decoding BitList.\nMissing sentinel bit.\n"}
 
       len > max_size ->
-        {:error, "out of bounds"}
+        {:error,
+         "Invalid binary length while decoding BitList. \nExpected max_size: #{max_size}. Found: #{len}.\n"}
 
       true ->
         {:ok, decoded}
     end
   end
 
-  defp decode_bitlist(_bit_list, _max_size), do: {:error, "invalid bitlist"}
+  defp decode_bitvector(bit_vector, _size) when bit_size(bit_vector) == 0 do
+    {:error, "Invalid binary value while decoding BitVector.\nEmpty binary found.\n"}
+  end
 
   defp decode_bitvector(bit_vector, size) do
-    num_bytes = byte_size(bit_vector)
+    num_bits = bit_size(bit_vector)
+    num_bytes = div(num_bits, 8)
+    expected_bytes = max(1, div(size + 7, 8))
 
     cond do
-      bit_size(bit_vector) == 0 ->
-        {:error, "ExcessBits"}
+      num_bytes != expected_bytes ->
+        {:error,
+         "Invalid binary length while decoding BitVector. \nExpected: #{expected_bytes} bytes.\nFound: #{num_bytes} bytes."}
 
-      num_bytes != max(1, div(size + 7, 8)) ->
-        {:error, "InvalidByteCount"}
+      match?(
+        <<_first::binary-size(num_bytes - 1), 0::size(8 - rem(size, 8) &&& 7), _rest::bitstring>>,
+        bit_vector
+      ) ->
+        {:ok, BitVector.new(bit_vector, size)}
 
       true ->
-        case bit_vector do
-          # Padding bits are clear
-          <<_first::binary-size(num_bytes - 1), 0::size(8 - rem(size, 8) &&& 7),
-            _rest::bitstring>> ->
-            {:ok, BitVector.new(bit_vector, size)}
-
-          _else ->
-            {:error, "ExcessBits"}
-        end
+        {:error, "Invalid binary length while decoding BitVector. \nExpected size: #{size}.\n"}
```

Comment:
We can change this for a `case` construct by constraining the binary matching a bit more:

```elixir
padding_bits = rem(8 - rem(size, 8), 8)

case bit_vector do
  <<_first::binary-size(num_bytes - 1), 0::size(padding_bits), _rest::size(8 - padding_bits)>> ->
    {:ok, BitVector.new(bit_vector, size)}
  # We can add some other error clauses...
  _ ->
    {:error, "Invalid binary length while decoding BitVector. \nExpected size: #{size}.\n"}
end
```